### DISCUSSION TITLE
Lots of TTS examples, what about Speech to Text?

### DISCUSSION THREAD

**Discussion by MaxMaeder** [pschatzmann/arduino-audio-tools#1269] [Q&A] [answered]
I see a lot of examples of text-to-speech implementations, however I was wondering if anyone had experience implementing speech-to-text using this library and a API like the Azure Speech Service.

As for my specific hardware, I'm trying to use a I2S microphone and an esp32 with WiFi

**Comment by pschatzmann** [comment]
I have never tried, but I don't see any reason why this should not work...

**Reply by MaxMaeder** [reply]
For these APIs I need to POST the audio data as a binary wav, is there a way for me to `Stream` this data to a HTTP post endpoint in Arduino?
[docs](https://learn.microsoft.com/en-us/azure/ai-services/speech-service/get-started-speech-to-text?tabs=macos%2Cterminal&pivots=programming-language-rest#recognize-speech-from-a-file)

**Comment by pschatzmann** [comment, marked as answer]
Sure: have a look at the [HttpRequest class](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_http_request.html#a16e3686c0aee7323b2300d3b4e5f3480)

**Reply by MaxMaeder** [reply]
Oh perfect!

**Comment by kaloprojects** [comment]
Great thread, thank you both!. @MaxMaeder : that's exactly what i am missing to my use case too!.

Background (standalone ESP32 OpenAI chatbot): I'm also using an ESP32 wifi connected (Arduino IDE) .. entering a user input via Serial .. sending as request to OpenAI API (POST message) and receiving and printing out show response. Meanwhile also implemented a Open AI Text-To-Speech (works great, using schreibfaul1 [audio.openai_speech()](https://github.com/schreibfaul1/ESP32-audioI2S/blob/master/examples/OpenAI_Speech/src/main.cpp)), so i can listen to the AI answer via I2S amplifier (MAX98357) ....
... _BUT what i totally miss is a connection to any Speech-To-Text API_ via my connected I2S microphone INMP441 (having any STT API call would allow me to replace my keyboard with my voice and having a standalone solution).

I might not have the skills for implementing the STT by myself :|  (even with @pschatzmann great link) .. so i cross my finger you can have progress to share ? :) 

What i wished is something like (any code snippets welcome!), using @pschatzmann's library:
- recording an audio (e.g. 3-10 sec, as long i press a btn) with my I2S microhone (in case a I2S SD card storage needed i could add)
- sending this wav file/buffer (or even as stream ?, would be best) to an STT Service (via POST message) and receiving text transcription from API. 

I would love most this [STT API from Open AI](https://platform.openai.com/docs/guides/speech-to-text)), or alternativelly [Deepgram](https://deepgram.com/product/speech-to-text), both seem very powerful, not expensive, multi languages detection, and most user (me included) have an account already, so no complex Azure registration or costs needed. But i accept any vendor (e.g. Azure), as i currently have no STT idea at all.

@MaxMaeder : do you think you can share some first code examples how to call such an SST API (with pschatzmann help above) ?

**Reply by arduibag** [reply]
> @MaxMaeder : do you think you can share some first code examples how to call such an SST API (with pschatzmann help above) ?
It would be great!

**Reply by MaxMaeder** [reply]
Made a demo, see below

**Comment by MaxMaeder** [comment]
I made a little Assistant demo: uses Azure SST/TTS + OpenAI API.

https://github.com/MaxMaeder/arduino-audio-tools-assistant
